packages/interaction/src/InteractionManager.js

```diff
@@ -1130,6 +1154,22 @@ export default class InteractionManager extends EventEmitter
             }
         }
 
+        const delayedEvents = this.delayedEvents;
```

Comment:
You can do this without creating a new array.

```suggestion
        const { delayedEvents } = this;
        const { length } = delayedEvents;
        
        if (delayed && length)
        {
                for (let i = 0; i < length / 3; i += 3)
                {
                       this.dispatchEvent(
                              delayedEvents[i],
                              delayedEvents[i+1],
                              delayedEvents[i+2]
                       );
                }
                delayedEvents.length = 0;
        }
```

Reply:
I don't know whether something would make this safe further up the chain, but iterating over callbacks without using that new-array pattern is pretty much never a good idea.

Imagine one of the callbacks causes another event to be processed, if it adds another delayed event we'll iterate over everything twice.

Reply:
Rule: never iterate over a collection that can mutate during the iteration. Unless you really REALLY know what you're doing.
Corollary: if you call unknown user code during an iteration, always assume that the user code will mutate your collection.

Reply:
Then slice/clone the array or iterative over it backwards to prevent issues. This is a reason why we needs unit tests.

Reply:
If we slice/clone it, we still need to clear the original array. That's just twice the work.

If we really care about making a code path that's not really all that hot (it's only taken on mouseout) as fast as possible, setting ``.length = 0`` is probably not the way to go about it.

If we do care about event order, then we should probably keep it the way it is. If we *don't* care about event order and don't mind going backwards through the array, then ``.pop()`` is probably the fastest way to safely go about this.

![5528](https://user-images.githubusercontent.com/1431530/63185024-37111700-c059-11e9-9db4-6df1a67bb5df.png)

https://jsperf.com/sum-and-clear-an-array/22

Reply:
Pop is good. I’m glad you’re being performance sensitive here. Interaction can be a bottleneck for some. Want to make sure we’re doing this right!

Reply:
Can you demonstrate or make a test for where order matters? I’m assuming in this case it’s not that relevant. What’s more relevant is the registered event handlers. And the eventemitter handler queue is handled separately.

Reply:
Ping @laino ☝️

Reply:
> Can you demonstrate or make a test for where order matters in delayedEvents?

Normally events are dispatched in the order of "children before parent elements", which in my book is a nice promise to the user and reversing the order of events for pointerout/mouseout breaks that.

In any case this codepath isn't near hot enough. That array is maybe going to contain 5 elements at most when the user actually left some element with their mouse, which maybe happens a dozen times a second tops in a very busy scene.